packages/walletconnect/src/index.ts

```diff
@@ -73,81 +55,81 @@ export class WalletConnect extends Connector {
-          Number(chainId),
-          await getBestUrl(this.rpc[Number(chainId)], this.timeout),
-        ]
+    if (this.provider) {
+      await this.deactivate()
+    }
+    
+    const rpcMap = this.rpcMap
+    const resolvedRpcMap = rpcMap ? Object.fromEntries(
+      await Promise.all(
+        Object.entries(rpcMap).map(
+          async ([chainId, map]): Promise<[string, string]> => [
+            `${chainId}`,
+            await getBestUrl(map, this.timeout),
+          ]
+        )
       )
-    ).then((results) =>
-      results.reduce<{ [chainId: number]: string }>((accumulator, [chainId, url]) => {
-        accumulator[chainId] = url
-        return accumulator
-      }, {})
-    )
-
+    ) : undefined
+    
     return (this.eagerConnection = import('@walletconnect/ethereum-provider').then(async (m) => {
-      this.provider = new m.default({
+      const provider = this.provider = await m.default.init({
         ...this.options,
-        chainId,
-        rpc: await rpc,
+        rpcMap: resolvedRpcMap,
       }) as unknown as MockWalletConnectProvider
 
-      this.provider.on('disconnect', this.disconnectListener)
-      this.provider.on('chainChanged', this.chainChangedListener)
-      this.provider.on('accountsChanged', this.accountsChangedListener)
-      this.provider.connector.on('display_uri', this.URIListener)
+      provider.on('disconnect', this.disconnectListener)
+      provider.on('chainChanged', this.chainChangedListener)
+      provider.on('accountsChanged', this.accountsChangedListener)
+      provider.on('display_uri', this.URIListener)
+      
+      return provider
     }))
   }
 
   /** {@inheritdoc Connector.connectEagerly} */
   public async connectEagerly(): Promise<void> {
     const cancelActivation = this.actions.startActivation()
 
+    /**
+     * WalletConnect automatically restores the session on `init` and updates the `accounts` and `chainId` properties.
+     * Therefore, it is not neccessary to call `eth_accounts` and `eth_chainId` to get the accounts and chainId.
+     */
     try {
-      await this.isomorphicInitialize()
-      if (!this.provider?.connected) throw Error('No existing connection')
-
-      // for walletconnect, we always use sequential instead of parallel fetches because otherwise
-      // chainId defaults to 1 even if the connecting wallet isn't on mainnet
-      const accounts = await this.provider?.request<string[]>({ method: 'eth_accounts' })
-      if (!accounts.length) throw new Error('No accounts returned')
-      const chainId = await this.provider
-        .request<string | number>({ method: 'eth_chainId' })
-        .then((chainId) => parseChainId(chainId))
-
-      this.actions.update({ chainId, accounts })
-    } catch (error) {
+      const provider = await this.isomorphicInitialize()
+      if (!provider.accounts.length) {
+        throw new Error('No accounts returned')
+      }
+      this.actions.update({ accounts: provider.accounts, chainId: provider.chainId })
```

Comment:
is there any case where you would have accounts but not a chainId?

Reply:
@cmcewen I have actually updated this code to be a bit more comprehensive, it now checks on `session`, which I have confirmed with the team to be the right way to do it.

Reply:
is there documentation that you could link to in an inline comment that states this?

Reply:
I think including a high-level explanation would be good, and a link to a specific line of code would be better; the goal being to make sure that a future maintainer (or someone making a fix) won't have to re-investigate the WC sourcecode to figure out _why_ this was the way you decided to do this.

Reply:
Done